### PR TITLE
Add Resque instrumentation

### LINKED ISSUE #75: Add Resque instrumentation and error reporting
As Resque (by default) forks the process around every job this was a bit tricky. We want this eventually but for now it's put off until after 1.0.

### PR DESCRIPTION
**Pull request by estolfo**
Closes #75

### CODE REVIEW

File: CHANGELOG.asciidoc [hunk 1]
```diff
@@ -44,6 +44,7 @@ endif::[]
 
 - Support for gRPC using the `grpc` gem (Experimental) {pull}669[#669]
 - Add `span.context.destination.address` and `span.context.destination.port` when available. {pull}722[#722]
+- Add support for instrumenting `Resque`. {pull}729[#729]
```
Comment (mikker) [inline CHANGELOG.asciidoc]:
```suggestion
- Add support for Resque. {pull}729[#729]
```

Not really _code_.

File: docs/api.asciidoc [hunk 2]
```diff
@@ -42,6 +42,16 @@ If you're not using Rails, see <<getting-started-rack,Getting started with Rack>
 Stop the currently running agent. Use this inside `at_exit` in your
 <<getting-started-rack,Rack app>> to gracefully shut down.
 
+[float]
+[[api-agent-restart]]
+==== `ElasticAPM.restart`
+
+If the agent is already running, this method will stop and start the agent.
```
Comment (mikker) [inline docs/api.asciidoc]:
```suggestion
If the agent is already running, this method will stop and start the agent.

```

File: docs/api.asciidoc [hunk 3]
```diff
@@ -42,6 +42,16 @@ If you're not using Rails, see <<getting-started-rack,Getting started with Rack>
...
+If the agent is not already running, this method will start the agent.
```
Comment (mikker) [inline docs/api.asciidoc]:
```suggestion
If the agent is not already running, this method will start the agent.

```

File: docs/supported-technologies.asciidoc [hunk 4]
```diff
@@ -86,6 +86,33 @@ We automatically instrument background processing using:
...
 - Sneakers (v2.12.0+) (Experimental, see https://github.com/elastic/apm-agent-ruby/pull/676[#676])
+- Resque (v2.0.0+)
+
+[float]
+[[supported-technologies-resque]]
+=== Resque
+
+You should require `elastic_apm/resque` in the Rakefile where the Resque rake tasks are required.
+Doing this will set up a hook to automatically start the ElasticAPM agent after Resque forks for the
+first time.
```
Comment (mikker) [inline docs/supported-technologies.asciidoc]:
```suggestion
```

I don't think we need to explain _why_ here.

File: docs/supported-technologies.asciidoc [hunk 5]
```diff
@@ -86,6 +86,33 @@ We automatically instrument background processing using:
...
+For example, in your `Rakefile`:
+
+[source,ruby]
+----
+require 'resque'
+require 'elastic_apm'
+require 'elastic_apm/resque'
+----
+
+When you start Resque, you should see a series of messages like the following in the Resque logs:
```
Comment (mikker) [inline docs/supported-technologies.asciidoc]: 👌

File: lib/elastic_apm.rb [hunk 6]
```diff
@@ -46,6 +46,16 @@ def stop
       Agent.stop
     end
 
+    # Restarts the ElasticAPM Agent using the same config or a new
+    # one, if it is provided.
+    # Starts the agent if it is not running.
+    # Stop ands starts the agent if it is running.
```
Comment (mikker) [inline lib/elastic_apm.rb]:
```suggestion
    # Stops and starts the agent if it is running.
```

File: spec/elastic_apm/resque_spec.rb [hunk 7]
```diff
@@ -0,0 +1,11 @@
+# frozen_string_literal: true
+
+require 'resque'
+
+RSpec.describe Resque do
+  it 'defines a before_first_fork hook' do
+    hooks_count = ::Resque.instance_variable_get(:@hooks).size
+    require 'elastic_apm/resque'
+    expect(::Resque.instance_variable_get(:@hooks).size).to eq(hooks_count + 1)
+  end
```
Comment (mikker) [inline spec/elastic_apm/resque_spec.rb]: Not sure I see the need for this spec.

File: spec/elastic_apm_spec.rb [hunk 8]
```diff
@@ -13,6 +13,22 @@
       ElasticAPM.stop
       expect(ElasticAPM::Agent).to_not be_running
     end
+
+    it 'restarts the agent', :mock_intake do
```
Comment (mikker) [inline spec/elastic_apm_spec.rb]:
```ruby
describe ".restart" do
  it "starts if stopped" # ...
  it "restarts if running"
end
```

Might provide more helpful error messages.

File: docs/supported-technologies.asciidoc (hunk 4 above)
Comment (estolfo) [reply docs/supported-technologies.asciidoc]: I think we should tell users that the agent will automatically start upon at the first fork though.

File: spec/elastic_apm/resque_spec.rb (hunk 7 above)
Comment (estolfo) [reply spec/elastic_apm/resque_spec.rb]: I just wasn't comfortable not testing the ElasticAPM::Resque module at all (or only indirectly in the resque spy spec). Do you think we should just remove this spec entirely?

File: docs/supported-technologies.asciidoc (hunk 4 above)
Comment (mikker) [reply docs/supported-technologies.asciidoc]: We're already saying _Do this and Resque works_. I don't feel like we need to explain _why_ it does or _how_ it does so. It's all there in the code if anyone's curious.

Comment (estolfo) [reply docs/supported-technologies.asciidoc]: Also, they need to know that the agent will not automatically start if they don't include `elastic_apm/resque` in the `Rakefile`.

Comment (mikker) [reply docs/supported-technologies.asciidoc]: The instructions are needed, not disagreeing.

```
To make the agent work with Resque, you need to require `elastic_apm/resque` before you boot your Resque worker process.

For example in your `Rakefile`:

# ... blah blah
```

I think something like that is sufficient. People don't notice the details anyway, so better get straight to the code example 😅

Comment (estolfo) [reply docs/supported-technologies.asciidoc]: ok👍🏼

File: spec/elastic_apm/resque_spec.rb (hunk 7 above)
Comment (mikker) [reply spec/elastic_apm/resque_spec.rb]: Yes. I think this is better tested at an integration(ish) level.

**Comment by estolfo** [comment]
Forgot about error handling 😬

File: lib/elastic_apm/spies/resque.rb [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+          ElasticAPM.restart
+        end
+      end
+
+      def install_perform_spy
+        ::Resque::Job.class_eval do
+          alias :perform_without_elastic_apm :perform
+
+          def perform
+            name = @payload && @payload['class'].name
```
Comment (estolfo) [inline lib/elastic_apm/spies/resque.rb]: Figured out a way to get the Job class name to add to the transaction.

Comment (mikker) [reply lib/elastic_apm/spies/resque.rb]: Wonderful!

**Pull request merged**